core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,263 @@
+        }
+
+        internal fun fromRangeTo(start: LocalDate, endInclusive: LocalDate) : LocalDateRange {
+            return LocalDateRange(start, endInclusive)
+        }
+    }
+}
+
+/**
+ * Returns the first [LocalDate] of the [LocalDateProgression].
+ *
+ * @throws NoSuchElementException if the progression is empty.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.firstAndLast
+ */
+public fun LocalDateProgression.first(): LocalDate {
+    if (isEmpty())
+        throw NoSuchElementException("Progression $this is empty.")
+    return this.first
+}
+
+/**
+ * Returns the last [LocalDate] of the [LocalDateProgression].
+ *
+ * @throws NoSuchElementException if the progression is empty.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.firstAndLast
+ */
+public fun LocalDateProgression.last(): LocalDate {
+    if (isEmpty())
+        throw NoSuchElementException("Progression $this is empty.")
+    return this.last
+}
+
+/**
+ * Returns the first [LocalDate] of the [LocalDateProgression], or null if the progression is empty.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.firstAndLast
+ */
+public fun LocalDateProgression.firstOrNull(): LocalDate? = if (isEmpty()) null else this.first
+
+/**
+ * Returns the last [LocalDate] of the [LocalDateProgression], or null if the progression is empty.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.firstAndLast
+ */
+public fun LocalDateProgression.lastOrNull(): LocalDate? = if (isEmpty()) null else this.last
+
+/**
+ * Returns a reversed [LocalDateProgression], i.e. one that goes from [last] to [first].
+ * The sign of the step is switched, in order to reverse the directionality of the progression.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.reversedProgression
+ */
+public fun LocalDateProgression.reversed(): LocalDateProgression = LocalDateProgression(longProgression.reversed())
+
+/**
+ * Returns a [LocalDateProgression] with the same start and end, but a changed step value.
+ *
+ * **Pitfall**: the value parameter represents the magnitude of the step, not the direction, and therefore must be positive.
+ * Its sign will be matched to the sign of the existing step, in order to maintain the directionality of the progression.
+ * If you wish to switch the direction of the progression, use [LocalDateProgression.reversed]
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.progressionWithStep
+ */
+public fun LocalDateProgression.step(value: Int, unit: DateTimeUnit.DayBased) : LocalDateProgression = step(value.toLong(), unit)
+
+/**
+ * Returns a [LocalDateProgression] with the same start and end, but a changed step value.
+ *
+ * **Pitfall**: the value parameter represents the magnitude of the step, not the direction, and therefore must be positive.
+ * Its sign will be matched to the sign of the existing step, in order to maintain the directionality of the progression.
```

Comment:
> directionality

I don't see a suitable meaning in https://www.merriam-webster.com/dictionary/directionality
Maybe just "direction"?

Reply:
This feels like one of those cases where I have an intuitive feel of the nuances of English, but can't point to a hard and fast rule. A progression has directionality, as well as having _a_ direction. I think in this case either `directionality` or `direction` is valid, so I'll defer to whichever you prefer.

Reply:
My (non-native) intuitive feel of English tells me that ranges as a concept have directionality, but a specific range has a direction. Also, if both words can be used here, I'd stick to the simpler one so as not to distract the reader. API references should strive to be boring to read, I think.